connection.go

```diff
@@ -1094,13 +1140,30 @@ func (conn *Connection) send(req Request, streamId uint64) *Future {
 		default:
 		}
 	}
+
+	if conn.shutdownWatcher != nil {
+		atomic.AddUint32(&(conn.requestCnt), uint32(1))
+		go conn.gracefulWait(fut)
+	}
+
 	conn.putFuture(fut, req, streamId)
+
 	if req.Ctx() != nil {
 		go conn.contextWatchdog(fut, req.Ctx())
 	}
+
 	return fut
 }
 
+func (conn *Connection) gracefulWait(fut *Future) {
+	<-fut.done
+	// This is a real advice from Go documentation
+	// about how to decrement atomic uint32.
+	// https://pkg.go.dev/sync/atomic#AddUint32
+	atomic.AddUint32(&(conn.requestCnt), ^uint32(0))
+	conn.cond.Broadcast()
+}
```

Comment:
1. Running a goroutine for every request looks too expensive for such a simple case (we can just decrement the counter in markDone (and a couple more places?).
2. For the counter, you can just use int64 instead of uint32.
3. We don't need to broadcast every time, it's enough:
```Go
    if atomic.AddUint32(&conn.RequestCnt, -1) == 0 {
        conn.cond.Broadcast()
    }
```

Reply:
Reworked